File: app/src/main/java/com/example/cinemates/view/ui/details/actor/ActorDetailsFragment.java
package com.example.cinemates.view.ui.details.actor;


import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;

import com.example.cinemates.R;
import com.example.cinemates.adapter.ItemsRecyclerViewAdapter;
import com.example.cinemates.databinding.FragmentActorDetailsBinding;
import com.example.cinemates.model.data.Actor;
import com.example.cinemates.model.data.Movie;
import com.example.cinemates.model.data.Person;
import com.example.cinemates.util.ViewSize;
import com.example.cinemates.view.ui.ActorDetailsFragmentArgs;
import com.example.cinemates.view.viewmodel.DbViewModel;
import com.example.cinemates.view.viewmodel.MovieViewModel;

import java.util.List;

public class ActorDetailsFragment extends Fragment {

    private FragmentActorDetailsBinding mBinding;
    private ItemsRecyclerViewAdapter<Movie> mAdapter;
    private MovieViewModel viewModel;
    private DbViewModel dbViewModel;
    private Person person;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        dbViewModel = new ViewModelProvider(getActivity()).get(DbViewModel.class);
        person = ActorDetailsFragmentArgs.fromBundle(getArguments()).getPerson();
        viewModel = new ViewModelProvider(getActivity()).get(MovieViewModel.class);
        mAdapter = new ItemsRecyclerViewAdapter<>(ViewSize.SMALL);

    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        mBinding = FragmentActorDetailsBinding.inflate(
                inflater,
                container,
                false
        );

        return mBinding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        if (dbViewModel.getPerson(person) != null)
            mBinding.fab.setImageDrawable(getActivity().getDrawable(R.drawable.ic_baseline_favorite_24));
        else
            mBinding.fab.setImageDrawable(getActivity().getDrawable(R.drawable.ic_baseline_favorite_border_24));

        mBinding.recyclerView.setAdapter(mAdapter);

        viewModel.getActor().observe(getViewLifecycleOwner(), new Observer<Actor>() {
            @Override
            public void onChanged(Actor actor) {
                mBinding.setActor(actor);
            }
        });
        viewModel.getMoviesByActor().observe(getViewLifecycleOwner(), new Observer<List<Movie>>() {
            @Override
            public void onChanged(List<Movie> movies) {
                mAdapter.addItems(movies);
            }
        });

        viewModel.getMoviesByActor(String.valueOf(person.getId()));
        viewModel.getActorDetails(person.getId());

        mBinding.toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Navigation.findNavController(view).navigateUp();
            }
        });

        mBinding.fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                person.setFavorite();
                if (dbViewModel.getPerson(person) != null) {//it was already into favorite
                    dbViewModel.delete(person);//should delete it
                    mBinding.fab.setImageDrawable(getActivity().getDrawable(R.drawable.ic_baseline_favorite_border_24));
                    Toast.makeText(getContext(), "Removed from favorite", Toast.LENGTH_SHORT).show();
                } else {
                    dbViewModel.insert(person);
                    mBinding.fab.setImageDrawable(getActivity().getDrawable(R.drawable.ic_baseline_favorite_24));
                    Toast.makeText(getContext(), "Added to favorite", Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mBinding = null;
    }


}

File: app/src/main/java/com/example/cinemates/view/ui/details/actor/ActorDetailsFragment.kt
package com.example.cinemates.view.ui.details.actor

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnticipateOvershootInterpolator
import android.widget.Toast
import androidx.core.view.doOnPreDraw
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.interpolator.view.animation.FastOutSlowInInterpolator
import androidx.navigation.Navigation.findNavController
import androidx.navigation.fragment.navArgs
import com.example.cinemates.R
import com.example.cinemates.view.ui.adapter.ViewPagerAdapter
import com.example.cinemates.databinding.FragmentActorDetailsBinding
import com.example.cinemates.view.dbviewmodel.DbPersonViewModel
import com.example.cinemates.util.getLong
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.tabs.TabLayoutMediator
import com.google.android.material.transition.MaterialFadeThrough
import com.google.android.material.transition.MaterialSharedAxis

class ActorDetailsFragment : Fragment() {
    private var _binding: FragmentActorDetailsBinding? = null
    private val binding: FragmentActorDetailsBinding
        get() = _binding!!
    private val dbViewModel: DbPersonViewModel by activityViewModels()
    private val viewModel: ActorDetailsViewModel by activityViewModels()
    private val args: ActorDetailsFragmentArgs by navArgs()
    private lateinit var viewPagerAdapter: ViewPagerAdapter
    private lateinit var actorAboutFragment: ActorAboutFragment
    private lateinit var actorMoviesFragment: ActorMoviesFragment


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enterTransition = MaterialSharedAxis(MaterialSharedAxis.Z, true).apply {
            interpolator = AnticipateOvershootInterpolator()
            duration = resources.getLong(R.integer.material_motion_duration_medium_2)
        }
        reenterTransition = MaterialSharedAxis(MaterialSharedAxis.Z, false).apply {
            interpolator = AnticipateOvershootInterpolator()
            duration = resources.getLong(R.integer.material_motion_duration_long_2)
        }
        returnTransition = MaterialFadeThrough()
        exitTransition = MaterialFadeThrough().apply {
            interpolator = FastOutSlowInInterpolator()
            duration = resources.getLong(R.integer.material_motion_duration_short_1)
        }
        actorAboutFragment = ActorAboutFragment()
        actorMoviesFragment = ActorMoviesFragment()
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        _binding = FragmentActorDetailsBinding.inflate(
            inflater,
            container,
            false
        )

        initializeViewPager()
        return binding.root
    }

    private fun initializeViewPager() {
        viewPagerAdapter = ViewPagerAdapter(childFragmentManager, lifecycle)
        viewPagerAdapter.addFragment(actorAboutFragment)
        viewPagerAdapter.addFragment(actorMoviesFragment)
        binding.apply {
            viewPager.adapter = viewPagerAdapter
            TabLayoutMediator(tabLayout, viewPager) { tab, position ->
                when (position) {
                    0 -> tab.text = "About"
                    1 -> tab.text = "Movies"
                }
            }.attach()
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        postponeEnterTransition()
        view.doOnPreDraw { startPostponedEnterTransition() }
        viewModel.onDetailsFragmentReady(args.person.id)

        binding.apply {
            customizeFab(fab)
            fab.setOnClickListener {
                updateThisPerson(fab)
            }
            toolbar.setNavigationOnClickListener { view ->
                findNavController(view).navigateUp()
            }
            person = args.person

        }

    }

    private fun updateThisPerson(fab: FloatingActionButton) {
        if (dbViewModel.setAsFavorite(args.person)) {
            Toast.makeText(context, "Added to favorites", Toast.LENGTH_SHORT).show()
            fab.setImageResource(R.drawable.ic_favorite_filled)
        } else {
            Toast.makeText(context, "Removed from favorites", Toast.LENGTH_SHORT).show()
            fab.setImageResource(R.drawable.ic_favorite_border)
        }
    }

    private fun customizeFab(fab: FloatingActionButton) {
        if (dbViewModel.isMyFavoritePerson(args.person)) {
            fab.setImageResource(R.drawable.ic_favorite_filled)
        } else {
            fab.setImageResource(R.drawable.ic_favorite_border)
        }

    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
